File: pdf-rag/src/chunking/chunker.py
import hashlib
import os
from datetime import datetime, timezone
from typing import List, Dict, Any

from llama_index.core.node_parser import MarkdownNodeParser, SentenceSplitter
from llama_index.core import Document


class MarkdownChunker:
    def __init__(self, chunk_size: int = 512, chunk_overlap: int = 100):
        self.markdown_parser = MarkdownNodeParser()
        self.sentence_splitter = SentenceSplitter(
            chunk_size=chunk_size,
            chunk_overlap=chunk_overlap,
            separator=" ",
            paragraph_separator="\n\n",
        )

    def extract_headers_context(self, content: str, chunk_text: str) -> List[str]:
        lines = content.split("\n")
        chunk_start = content.find(chunk_text[:100])

        headers = []
        current_headers = []

        for i, line in enumerate(lines):
            if line.strip().startswith("#"):
                level = len(line) - len(line.lstrip("#"))
                header_text = line.lstrip("# ").strip()

                current_headers = [h for h in current_headers if h[0] < level]
                current_headers.append((level, header_text))

                line_pos = sum(len(l) + 1 for l in lines[:i])
                if chunk_start >= 0 and line_pos >= chunk_start:
                    headers = [h[1] for h in current_headers]
                    break

        if not headers and current_headers:
            headers = [h[1] for h in current_headers]

        return headers

    def chunk_markdown_with_pages(self, pages_data: List[Dict[str, Any]], source_file: str) -> List[Dict[str, Any]]:
        print(f"Chunking content from {len(pages_data)} pages...")

        all_chunks = []
        doc_hash = hashlib.md5("".join([p["content"] for p in pages_data]).encode()).hexdigest()

        for page_data in pages_data:
            page_content = page_data["content"]
            page_num = page_data["page_number"]
            total_pages = page_data["total_pages"]

            print(f"  Processing page {page_num}/{total_pages}...")

            doc = Document(text=page_content)
            markdown_nodes = self.markdown_parser.get_nodes_from_documents([doc])

            page_chunks = []
            for node in markdown_nodes:
                sentence_nodes = self.sentence_splitter.get_nodes_from_documents([Document(text=node.text)])
                page_chunks.extend(sentence_nodes)

            for i, chunk in enumerate(page_chunks):
                headers = self.extract_headers_context(page_content, chunk.text)

                overlap_content = ""
                if i > 0:
                    overlap_content += page_chunks[i - 1].text[-100:]
                elif page_num > 1 and all_chunks:
                    overlap_content += all_chunks[-1]["content"][-100:]

                if i < len(page_chunks) - 1:
                    overlap_content += page_chunks[i + 1].text[:100]

                chunk_data = {
                    "content": chunk.text,
                    "title": os.path.splitext(os.path.basename(source_file))[0],
                    "sourceFile": source_file,
                    "chunkIndex": len(all_chunks),
                    "chunkSize": len(chunk.text),
                    "precedingHeaders": headers,
                    "overlapContent": overlap_content,
                    "documentHash": doc_hash,
                    "processedAt": datetime.now(timezone.utc),
                    "pageNumber": page_num,
                    "pageChunkIndex": i,
                    "totalPages": total_pages,
                    "chunksInPage": len(page_chunks),
                    "chunkMethod": "page_aware_markdown_sentence_split",
                    "pageCharCount": len(page_content),
                }
                all_chunks.append(chunk_data)

            print(f"    ✅ Page {page_num}: {len(page_chunks)} chunks")

        print(f"✅ Created {len(all_chunks)} total chunks from {len(pages_data)} pages")
        return all_chunks

File: pdf-rag-ts/src/chunking/chunker.ts
import * as crypto from 'crypto';
import * as path from 'path';
import { Document, MarkdownNodeParser, SentenceSplitter } from 'llamaindex';
import type { PageData } from '../ingestion/parser.js';
import { parallelMap } from '../utils/helpers.js';

export type ChunkingStrategy = 'STANDARD' | 'HIERARCHICAL' | 'CONTEXTUAL' | 'FULL_PAGE';

export interface ChunkData {
    content: string;
    title: string;
    sourceFile: string;
    chunkIndex: number;
    chunkSize: number;
    precedingHeaders: string[];
    overlapContent: string;
    documentHash: string;
    processedAt: string;
    pageNumber: number;
    pageChunkIndex: number;
    totalPages: number;
    chunksInPage: number;
    chunkMethod: string;
    pageCharCount: number;
    parentContent?: string | undefined; // For HIERARCHICAL
    contextSummary?: string | undefined; // For CONTEXTUAL
}

export class MarkdownChunker {
    private markdownParser: MarkdownNodeParser;
    private standardSplitter: SentenceSplitter;
    private hierarchicalSplitter: SentenceSplitter;

    constructor(chunkSize: number = 512, chunkOverlap: number = 100) {
        this.markdownParser = new MarkdownNodeParser();
        this.standardSplitter = new SentenceSplitter({
            chunkSize: chunkSize,
            chunkOverlap: chunkOverlap,
        });
        this.hierarchicalSplitter = new SentenceSplitter({
            chunkSize: 200, // Small chunks for embedding
            chunkOverlap: 50,
        });
    }

    private extractHeadersContext(pageContent: string, chunkText: string, preParsedHeaders?: { level: number, text: string, pos: number }[]): string[] {
        const chunkStart = pageContent.indexOf(chunkText.substring(0, 100));
        if (chunkStart < 0) return [];

        let headers: { level: number, text: string, pos: number }[] = [];
        
        if (preParsedHeaders) {
            headers = preParsedHeaders;
        } else {
            const lines = pageContent.split('\n');
            let currentPos = 0;
            for (const line of lines) {
                if (line.trim().startsWith('#')) {
                    const match = line.match(/^(#+)\s*(.*)$/);
                    if (match && match[1] && match[2]) {
                        headers.push({ 
                            level: match[1].length, 
                            text: match[2].trim(),
                            pos: currentPos
                        });
                    }
                }
                currentPos += line.length + 1;
            }
        }

        const relevantHeaders: string[] = [];
        let currentLevel = 999;

        for (let i = headers.length - 1; i >= 0; i--) {
            const header = headers[i];
            if (header && header.pos <= chunkStart && header.level < currentLevel) {
                relevantHeaders.unshift(header.text);
                currentLevel = header.level;
                if (currentLevel === 1) break;
            }
        }

        return relevantHeaders;
    }

    async chunkMarkdownWithPages(
        pagesData: PageData[], 
        sourceFile: string, 
        strategy: ChunkingStrategy = 'STANDARD',
        summaryProvider?: (text: string) => Promise<string>
    ): Promise<ChunkData[]> {
        console.log(`Chunking content from ${pagesData.length} pages using strategy: ${strategy}...`);

        const allChunks: ChunkData[] = [];
        const fullContent = pagesData.map(p => p.content).join('');
        const docHash = crypto.createHash('md5').update(fullContent).digest('hex');

        // PRE-FETCH SUMMARIES IN PARALLEL (Concurrency limit 4)
        const pageSummaries = new Map<number, string>();
        if (strategy === 'CONTEXTUAL' && summaryProvider) {
            console.log(`  Fetching contextual summaries for ${pagesData.length} pages in parallel...`);
            await parallelMap(pagesData, async (page) => {
                const summary = await summaryProvider(page.content);
                pageSummaries.set(page.pageNumber, summary);
            }, 4);
            console.log(`  ✅ All summaries generated.`);
        }

        for (const pageData of pagesData) {
            const pageContent = pageData.content;
            const pageNum = pageData.pageNumber;
            const totalPages = pageData.totalPages;
            const pageSummary = pageSummaries.get(pageNum) || "";

            console.log(`  Processing page ${pageNum}/${totalPages}...`);

            // PRE-PARSE HEADERS FOR THE PAGE
            const pageHeaders: { level: number, text: string, pos: number }[] = [];
            const lines = pageContent.split('\n');
            let currentPos = 0;
            for (const line of lines) {
                if (line.trim().startsWith('#')) {
                    const match = line.match(/^(#+)\s*(.*)$/);
                    if (match && match[1] && match[2]) {
                        pageHeaders.push({ 
                            level: match[1].length, 
                            text: match[2].trim(),
                            pos: currentPos
                        });
                    }
                }
                currentPos += line.length + 1;
            }

            const doc = new Document({ text: pageContent });
            const markdownNodes = await this.markdownParser.getNodesFromDocuments([doc]);

            const pageChunksNodes: any[] = [];
            const parentMap = new Map<any, string>();

            for (const node of markdownNodes) {
                if (strategy === 'HIERARCHICAL') {
                    const childNodes = await this.hierarchicalSplitter.getNodesFromDocuments([new Document({ text: node.text })]);
                    for (const child of childNodes) {
                        pageChunksNodes.push(child);
                        parentMap.set(child, node.text); 
                    }
                } else if (strategy === 'FULL_PAGE') {
                    pageChunksNodes.push(node);
                } else {
                    const sentenceNodes = await this.standardSplitter.getNodesFromDocuments([new Document({ text: node.text })]);
                    pageChunksNodes.push(...sentenceNodes);
                }
            }

            for (let i = 0; i < pageChunksNodes.length; i++) {
                const chunk = pageChunksNodes[i];
                const headers = this.extractHeadersContext(pageContent, chunk.text, pageHeaders);

                let overlapContent = "";
                if (i > 0) {
                    const prevChunk = pageChunksNodes[i - 1];
                    if (prevChunk) overlapContent += prevChunk.text.slice(-100);
                } else if (pageNum > 1 && allChunks.length > 0) {
                    const lastChunk = allChunks[allChunks.length - 1];
                    if (lastChunk) overlapContent += lastChunk.content.slice(-100);
                }

                if (i < pageChunksNodes.length - 1) {
                    const nextChunk = pageChunksNodes[i + 1];
                    if (nextChunk) overlapContent += nextChunk.text.slice(0, 100);
                }

                let content = chunk.text;
                if (strategy === 'CONTEXTUAL' && pageSummary) {
                    const contextPrefix = `[Context: ${headers.join(' > ')}${pageSummary ? ' | Page Summary: ' + pageSummary : ''}]\n`;
                    content = contextPrefix + content;
                }

                const chunkData: ChunkData = {
                    content: content,
                    title: path.parse(path.basename(sourceFile)).name,
                    sourceFile: sourceFile,
                    chunkIndex: allChunks.length,
                    chunkSize: content.length,
                    precedingHeaders: headers,
                    overlapContent: overlapContent,
                    documentHash: docHash,
                    processedAt: new Date().toISOString(),
                    pageNumber: pageNum,
                    pageChunkIndex: i,
                    totalPages: totalPages,
                    chunksInPage: pageChunksNodes.length,
                    chunkMethod: `page_aware_${strategy.toLowerCase()}`,
                    pageCharCount: pageContent.length,
                };
                
                if (parentMap.has(chunk)) {
                    chunkData.parentContent = parentMap.get(chunk);
                }
                if (pageSummary) {
                    chunkData.contextSummary = pageSummary;
                }

                allChunks.push(chunkData);
            }

            console.log(`    ✅ Page ${pageNum}: ${pageChunksNodes.length} chunks`);
        }

        console.log(`✅ Created ${allChunks.length} total chunks from ${pagesData.length} pages`);
        return allChunks;
    }
}
